src/synapse/api.py

```diff
@@ -129,23 +141,85 @@ def _generate_mac(
+
+    Args:
+        api_url: URL to be requested.
+        retry: call URL with a retry. Default is False.
+
+    Raises:
+        NetworkError: if there was an error fetching the api_url.
+
+    Returns:
+        Response from calling the URL.
+    """
+    try:
+        session = requests.Session()
+        retries = Retry(
+            total=3,
+            backoff_factor=3,
+        )
+        if retry:
+            session.mount("http://", HTTPAdapter(max_retries=retries))
+        res = session.get(api_url, timeout=10)
+        res.raise_for_status()
+        session.close()
+    except (requests.exceptions.ConnectionError, requests.exceptions.Timeout) as exc:
+        logger.exception("Failed to connect to %s: %r", api_url, exc)
+        raise NetworkError(f"Failed to connect to {api_url}.") from exc
+    except requests.exceptions.HTTPError as exc:
+        logger.exception("HTTP error from %s: %r", api_url, exc)
+        raise NetworkError(f"HTTP error from {api_url}.") from exc
+    return res
+
+
 def _get_nonce() -> str:
     """Get nonce.
 
     Returns:
         The nonce returned by Synapse API.
 
     Raises:
-        NetworkError: if there was an error fetching the nonce.
+        GetNonceError: if there was an error while reading nonce.
     """
+    res = _send_get_request(REGISTER_URL)
     try:
-        res = requests.get(REGISTER_URL, timeout=5)
-        res.raise_for_status()
-        return res.json()["nonce"]
-    except (
-        requests.exceptions.ConnectionError,
-        requests.exceptions.Timeout,
-        requests.exceptions.HTTPError,
-    ) as exc:
-        logger.error("Failed to request %s : %s", REGISTER_URL, exc)
-        raise NetworkError(f"Failed to request {REGISTER_URL}.") from exc
+        nonce = res.json()["nonce"]
+    except (requests.exceptions.JSONDecodeError, TypeError, KeyError) as exc:
+        logger.exception("Failed to decode nonce: %r. Received: %s", exc, res.text)
+        raise GetNonceError(str(exc)) from exc
+
+    return nonce
+
+
+def get_version() -> str:
+    """Get version.
+
+    Expected API output:
+    {
+        "server_version": "0.99.2rc1 (b=develop, abcdef123)",
+        "python_version": "3.7.8"
+    }
+
+    We're using retry here because after the config change, Synapse is restarted.
+
+    Returns:
+        The version returned by Synapse API.
+
+    Raises:
+        GetVersionError: if there was an error while reading version.
+    """
+    res = _send_get_request(VERSION_URL, retry=True)
+    try:
+        server_version = res.json()["server_version"]
+    except (requests.exceptions.JSONDecodeError, KeyError, TypeError) as exc:
+        logger.exception("Failed to decode version: %r. Received: %s", exc, res.text)
+        raise GetVersionError(str(exc)) from exc
+    version_match = re.search(SYNAPSE_VERSION_REGEX, server_version)
+    if not version_match:
+        # Exception not in docstring because is captured.
```

Comment:
Perhaps I'm missing something, where is the exception handled in this function please?

Reply:
The exception is actually raised, fixed.